cli/compose/template/template_test.go

```diff
@@ -172,3 +172,91 @@ func TestSubstituteWithCustomFunc(t *testing.T) {
 	_, err = SubstituteWith("ok ${NOTHERE}", defaultMapping, pattern, errIsMissing)
 	assert.Check(t, is.ErrorContains(err, "required variable"))
 }
+
+func TestExtractVariables(t *testing.T) {
+	testCases := []struct {
+		dict     map[string]interface{}
+		expected map[string]string
+	}{
+		{
+			dict:     map[string]interface{}{},
+			expected: map[string]string{},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "bar",
+			},
+			expected: map[string]string{},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "$bar",
+			},
+			expected: map[string]string{
+				"bar": "",
+			},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "${bar}",
+			},
+			expected: map[string]string{
+				"bar": "",
+			},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "${bar?:foo}",
+			},
+			expected: map[string]string{
+				"bar": "",
+			},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "${bar?foo}",
+			},
+			expected: map[string]string{
+				"bar": "",
+			},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "${bar:-foo}",
+			},
+			expected: map[string]string{
+				"bar": "foo",
+			},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "${bar-foo}",
+			},
+			expected: map[string]string{
+				"bar": "foo",
+			},
+		},
+		{
+			dict: map[string]interface{}{
+				"foo": "${bar:-foo}",
+				"bar": map[string]interface{}{
+					"foo": "${fruit:-banana}",
+					"bar": "vegetable",
+				},
+				"baz": []interface{}{
+					"foo",
+					"$toto",
+				},
+			},
+			expected: map[string]string{
+				"bar":   "foo",
+				"fruit": "banana",
+				"toto":  "",
+			},
+		},
+	}
+	for _, tc := range testCases {
```

Comment:
Could change this to subtests as a follow up?

Reply:
Yes sir ! :stuck_out_tongue_closed_eyes: